lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,145 @@
+-}
+subst :: Rule -> Pat -> Rule
+subst cst@(C _ t :<= _) ty = mapPat (go (getSubs t ty)) cst
+  where
+    go :: [(Text, Pat)] -> Pat -> Pat
+    go subs tty =
+      let noflip p1 p2 = uncurry subV p2 p1
+       in foldl' noflip tty subs
+
+{- Given two patterns (which are hopefully structurally similar), gather a list of all substitutions
+   from the PatVars in the first argument to the concrete types (hopefully!) in the second argument
+-}
+getSubs :: Pat -> Pat -> [(Text, Pat)] -- should be a set, whatever
+getSubs (VarP s) t = [(s, t)]
+getSubs (x :* xs) (x' :* xs') = getSubs x x' <> getSubs xs xs'
+getSubs (l := t) (l' := t') = getSubs l l' <> getSubs t t'
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+-- should be vastly more efficient than Data.List.Nub
+deduplicate :: Ord a => [a] -> [a]
+deduplicate = S.toList . S.fromList
+
+-- is the first pattern a substitution instance of the second
+isSubstitutionOf :: Pat -> Pat -> Bool
+isSubstitutionOf p1 p2 = matches p2 p1
+
+compareSpecificity :: Pat -> Pat -> Ordering
+compareSpecificity p1 p2
+  | p1
+      `isSubstitutionOf` p2
+      && p2
+      `isSubstitutionOf` p1 =
+      EQ
+  | p1 `isSubstitutionOf` p2 = LT
+  | otherwise = GT
+
+sortOnSpecificity :: Pat -> Class -> [Rule] -> [Rule]
+sortOnSpecificity p c ps =
+  sortBy (\a1 a2 -> compareSpecificity (ruleHeadPat a1) (ruleHeadPat a2)) $
+    filter matchPatAndClass ps
+  where
+    matchPatAndClass :: Rule -> Bool
+    matchPatAndClass r =
+      ruleHeadClass r == c
+        && ruleHeadPat r
+        `matches` p
+
+mostSpecificInstance :: Pat -> Class -> [Rule] -> Maybe Rule
+mostSpecificInstance p c ps = case sortOnSpecificity p c ps of
+  [] -> Nothing
+  (x : _) -> Just x
+
+{- Given a list of instances (the initial scope), determines whether we can derive
+   an instance of the Class argument for the Pat argument. A result of [] indicates that there are
+   no remaining subgoals and that the constraint has been solved.
+   NOTE: At the moment this handles superclasses differently than you might expect -
+         instead of assuming that the superclasses for all in-scope classes are defined,
+         we check that those constraints can be solved before affirmatively judging that the
+         target constraint has been solved. I *think* that makes sense in this context (whereas in Haskell
+         it doesn't b/c it's *impossible* to have `instance Foo X` if the definition of Foo is
+         `class Bar y => Foo y` without an `instance Bar X`)
+-}
+solve ::
+  [Rule] -> -- all instance rules in scope. WE ASSUME THESE HAVE ALREADY BEEN GENERATED, SOMEWHERE
```

Comment:
I would add turn this into a SolverContext and add all the rules there.

Reply:
Also can we please run the solver with a Reader context and a Writer one to collect the solver trace? That would be extremely useful!

Reply:
Er, yeah, we can do that.